### PR TITLE
Pattern mapping

### PR DESCRIPTION
**Pull request by ikbalkaya**
This PR implements pattern mapping by evaluating #{key} and #{topic} tokens in given configuration value.
* #{key} is only evaluated if key exists and is decodable to a UTF-8 string, If #{key} is given but a key has not been provided  by producer or a non-UTF8-decodable key is provided, Connect throws an `IllegalArgumentException`
* If #{topic} is given in a configuration value, it is always going to be evaluated as there cannot be any record without topics.  

Below is configuration values that current pattern evaluation is applied to
* Channel name 
* Cipher key
* Each channel param

### CODE REVIEW

File: src/main/java/com/ably/kafka/connect/ConfigValueEvaluator.java
Comment (KacperKluka) [inline src/main/java/com/ably/kafka/connect/ConfigValueEvaluator.java:42]: I'd like to propose to refactor this method to focus its flow around the `pattern` instead of the `key` :wink: Looking at this class tests I figured out that this method's result depends mostly on the provided pattern (given the topic and key values are okay), but I couldn't spot that right away when looking at the implementation :sweat_smile: We could change it to something like the code below :thinking: However, the current code works and I wouldn't like to block the PR so I'll let you decide if that's something we should change :+1: 
```java
    public String evaluate(SinkRecord record, String pattern) throws IllegalArgumentException {
        if (pattern == null) {
            return null;
        } else if (pattern.contains(KEY_TOKEN)) {
            final byte[] key = (byte[]) record.key();

            //we only want to evalutate UTF-8 encoded strings
            if (key != null && ByteArrayUtils.isUTF8Encoded(key)) {
                String keyString = new String(key, StandardCharsets.UTF_8);
                if (pattern.contains(TOPIC_TOKEN)){
                    return pattern.replace(KEY_TOKEN, keyString).replace(TOPIC_TOKEN, record.topic());
                } else {
                    return pattern.replace(KEY_TOKEN, keyString);
                }
            } else {
                throw new IllegalArgumentException("Key is null or not a string type but pattern contains ${key}");
            }
        } else if (pattern.contains(TOPIC_TOKEN)) {
            return pattern.replace(TOPIC_TOKEN, record.topic());
        } else {
            return pattern;
        }
    }
```

**Review by KacperKluka — CHANGES_REQUESTED** [review]
Overall looks good but we can improve it a bit more :grin:

File: src/main/java/com/ably/kafka/connect/ConfigValueEvaluator.java
Comment (ikbalkaya) [reply src/main/java/com/ably/kafka/connect/ConfigValueEvaluator.java:42]: A good point, but I would argue that current implementation is less verbose and more readable for reasons I will explain below.

Apart from early return, the decision block is this
```
if (keyString != null) {
            return pattern.replace(KEY_TOKEN, keyString).replace(TOPIC_TOKEN, record.topic());
        } else {
            return pattern.replace(TOPIC_TOKEN, record.topic());
        }
```
I prefer early returns for obvious cases where there can be some cases for the function that doesn't require further evaluation or processing (fail fast). And obviously it makes code less less branched. Just to summarise this method's structure
1. Early return based on given argument
2. Key evaluation - input for decision 
3. Decision

I somehow find reading code like this more clean and readable, might be a matter of taste.

File: src/test/java/com/ably/kafka/connect/DefaultChannelSinkMappingTest.java
Comment (ikbalkaya) [reply src/test/java/com/ably/kafka/connect/DefaultChannelSinkMappingTest.java]: That is a good point. The change is here https://github.com/ably/kafka-connect-ably/pull/43/commits/f2eaf019f8a43b31eabed98e191eb3b9f8b66245

File: src/main/java/com/ably/kafka/connect/ConfigValueEvaluator.java
Comment (ikbalkaya) [reply src/main/java/com/ably/kafka/connect/ConfigValueEvaluator.java]: Thanks for this suggestion. 
Here https://github.com/ably/kafka-connect-ably/pull/43/commits/372196a2fc0135400d4baa76c926fd4571d3408a
I applied your suggestion + a bit more explanation on key

**Review by KacperKluka — APPROVED** [review]
LGTM :+1:

**Review by lmars — APPROVED** [review]

**Pull request merged**
